### PR TITLE
[NFT Metadata Crawler] JSON Parser

### PR DESCRIPTION
**Pull request by just-in-chang**
### Description

Logic for JSON parser implementation

Fixes EI-251

### CODE REVIEW

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs [hunk 1]
```diff
@@ -9,7 +12,53 @@ impl JSONParser {
...
+     * Parses JSON from input URI.
+     */
+    async fn parse_json(
+        uri: String,
+    ) -> anyhow::Result<(Option<String>, Option<String>, Option<Value>)> {
+        let (mime, size) = get_uri_metadata(uri.clone()).await?;
+        if ImageFormat::from_mime_type(mime).is_some() {
+            error!("JSON parser received image URI, skipping");
+            return Ok((None, None, None));
+        } else if size > 5000000 {
```
Comment (larry-aptos) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs:35]: nit: const it at top.

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs [hunk 2]
```diff
@@ -9,7 +12,53 @@ impl JSONParser {
...
+            error!("JSON parser received large file, skipping");
+            return Ok((None, None, None));
+        }
+
+        for _ in 0..3 {
```
Comment (larry-aptos) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs:40]: nit: const at top for `3`

Comment (larry-aptos) [reply ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs:40]: also, let's do `backoff`

maybe this? https://docs.rs/backoff/latest/backoff/

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs [hunk 3]
```diff
@@ -9,7 +18,59 @@ impl JSONParser {
      * Parses JSON from input URI.
      * Returns the underlying raw image URI, raw animation URI, and JSON.
      */
-    pub async fn parse(_uri: String) -> (Option<String>, Option<String>, Option<Value>) {
-        todo!();
+    pub async fn parse(uri: String) -> (Option<String>, Option<String>, Option<Value>) {
+        match Self::parse_json(uri.clone()).await {
+            Ok(out) => out,
+            Err(e) => {
+                error!("Error parsing JSON: {}", e);
```
Comment (bowenyang007) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs:25]: use structured log: error=?e. Also add uri =uri

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs [hunk 4]
```diff
@@ -9,7 +18,59 @@ impl JSONParser {
...
+     */
+    async fn parse_json(
+        uri: String,
+    ) -> anyhow::Result<(Option<String>, Option<String>, Option<Value>)> {
+        let (mime, size) = get_uri_metadata(uri.clone()).await?;
+        if ImageFormat::from_mime_type(mime).is_some() {
+            error!("JSON parser received image URI, skipping");
+            return Ok((None, None, None));
+        } else if size > MAX_FILE_SIZE_BYTES {
+            error!("JSON parser received large file, skipping");
```
Comment (bowenyang007) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs:42]: all these errors should include uri = uri as context.

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs [hunk 5]
```diff
@@ -9,7 +18,59 @@ impl JSONParser {
...
+                    raw_image_uri,
+                    raw_animation_uri,
+                    Some(parsed_json),
+                ))
+            }
+            .boxed()
+        };
+
+        let backoff = ExponentialBackoff {
+            max_elapsed_time: Some(Duration::from_secs(MAX_RETRY_TIME_SECONDS)),
```
Comment (bowenyang007) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs:67]: so after it exceeds 15 sec it'll fail? how many times is that?

**Review by bowenyang007 — APPROVED** [review]
Some comments about logging.

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs [hunk 6]
```diff
@@ -9,7 +18,59 @@ impl JSONParser {
...
+    /**
+     * Parses JSON from input URI.
+     */
+    async fn parse_json(
+        uri: String,
+    ) -> anyhow::Result<(Option<String>, Option<String>, Option<Value>)> {
+        let (mime, size) = get_uri_metadata(uri.clone()).await?;
+        if ImageFormat::from_mime_type(mime).is_some() {
+            error!("JSON parser received image URI, skipping");
+            return Ok((None, None, None));
```
Comment (jjleng) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs:40]: why not returning error

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs [hunk 7]
```diff
@@ -9,7 +18,59 @@ impl JSONParser {
...
+        } else if size > MAX_FILE_SIZE_BYTES {
+            error!("JSON parser received large file, skipping");
+            return Ok((None, None, None));
+        }
+
+        let op = || {
+            async {
+                info!("Sending request for token_uri {}", uri);
+
+                let result = reqwest::get(&uri).await.map_err(Error::transient)?;
```
Comment (jjleng) [inline ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs:50]: do you want to distinguish the error codes? e.g. retry on 404 would likely fail again, right?

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs (hunk 5 above)
Comment (just-in-chang) [reply ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs:67]: Added to constants file, changed to 5 sec. Should be between 3-4 retries [documentation here](https://docs.rs/backoff/latest/backoff/)

File: ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs (hunk 7 above)
Comment (just-in-chang) [reply ecosystem/nft-metadata-crawler/nft-metadata-crawler-parser/src/utils/json_parser.rs]: planning on adding this as optimization in the future

**Pull request merged**